src/notify/notify.c

```diff
@@ -313,56 +337,227 @@ static int parse_argv(int argc, char *argv[]) {
+                        log_warning("Warning: %u more file descriptors passed than referenced with --fd=.", fdset_size(passed));
         }
 
-        if (have_env && arg_action == ACTION_BOOTED)
-                log_warning("Notify message specified along with --booted, ignoring.");
+        if (arg_action == ACTION_BOOTED && argc > optind)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "--booted takes no parameters, refusing.");
 
-        if (n_arg_env > 0) {
-                arg_env = strv_copy_n(argv + optind, n_arg_env);
-                if (!arg_env)
-                        return log_oom();
+        if (arg_action == ACTION_FORK && optind >= argc)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "--fork requires a command to be specified, refusing.");
+
+        return 1;
+}
+
+static int on_notify_socket(sd_event_source *s, int fd, unsigned event, void *userdata) {
+        PidRef *child = ASSERT_PTR(userdata);
+        int r;
+
+        assert(s);
+        assert(fd >= 0);
+
+        _cleanup_free_ char *text = NULL;
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        r = notify_recv(fd, &text, /* ret_ucred= */ NULL, &pidref);
+        if (r == -EAGAIN)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to receive notification message: %m");
+
+        if (!pidref_equal(child, &pidref)) {
+                log_warning("Received notification message from unexpected process " PID_FMT " (expected " PID_FMT "), ignoring.",
+                            pidref.pid, child->pid);
+                return 0;
         }
 
-        if (!fdset_isempty(passed))
-                log_warning("Warning: %u more file descriptors passed than referenced with --fd=.", fdset_size(passed));
+        const char *p = find_line_startswith(text, "READY=1");
+        if (!p || !IN_SET(*p, '\n', 0)) {
+                if (DEBUG_LOGGING) {
+                        _cleanup_free_ char *escaped = cescape(text);
+                        log_debug("Received notification message without READY=1, ignoring. (%s)", strna(escaped));
+                }
+        } else {
+                log_debug("Recieved READY=1, exiting.");
 
-        return 1;
+                r = sd_event_exit(sd_event_source_get_event(s), EXIT_SUCCESS);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to request event loop exit: %m");
+        }
+
+        return 0;
+}
+
+static int on_child(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        int r;
+
+        assert(s);
+        assert(si);
+
+        int ret;
+        if (si->si_code == CLD_EXITED) {
+                if (si->si_status != EXIT_SUCCESS)
+                        log_debug("Child failed with exit status %i.", si->si_status);
+                else
+                        log_debug("Child exited successfully");
+
+                ret = si->si_code;
```

Comment:
This is not zero, so we'll fail even if the child exited successfully, which I think is what we want if the child exits without sending READY=1, but this deserves a comment I think.

Reply:
oops, actually i wanted to propagate the original exit code here. will fix.